From a70c0cf81398a901a92db74c93c0cdf68987533e Mon Sep 17 00:00:00 2001
From: Abishek Ganapathy <quic_abishekg@quicinc.com>
Date: Tue, 2 Apr 2024 16:58:56 +0530
Subject: [PATCH] wifi: cfg80211: Remove WARN_ON during MLO STA down

When MLO STA is brought down, WARN_ON is observed due to scan being
requested for the partner vap. Following this, there is another
WARN_ON since the current BSS is still present.

Resolve this by replacing WARN_ON with appropriate print messages.

Change-Id: I501a3f64f4003a83eaa85b4536c47b6e4031af97
Signed-off-by: Abishek Ganapathy <quic_abishekg@quicinc.com>
---
 net/wireless/core.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/net/wireless/core.c b/net/wireless/core.c
index 563cfbe3237c..c30714f5e7e1 100644
--- a/net/wireless/core.c
+++ b/net/wireless/core.c
@@ -1245,7 +1245,8 @@ static void _cfg80211_unregister_wdev(struct wireless_dev *wdev,
 
 			curbss = wdev->links[link_id].client.current_bss;
 
-			if (WARN_ON(curbss)) {
+			if (curbss) {
+				printk("Freeing current BSS\n");
 				cfg80211_unhold_bss(curbss);
 				cfg80211_put_bss(wdev->wiphy, &curbss->pub);
 				wdev->links[link_id].client.current_bss = NULL;
@@ -1513,10 +1514,12 @@ static int cfg80211_netdev_notifier_call(struct notifier_block *nb,
 		wiphy_lock(&rdev->wiphy);
 		cfg80211_update_iface_num(rdev, wdev->iftype, -1);
 		if (rdev->scan_req && rdev->scan_req->wdev == wdev) {
-			if (WARN_ON(!rdev->scan_req->notified &&
+			if (!rdev->scan_req->notified &&
 				    (!rdev->int_scan_req ||
-				     !rdev->int_scan_req->notified)))
+				     !rdev->int_scan_req->notified)) {
+				printk("Aborting scan for device\n");
 				rdev->scan_req->info.aborted = true;
+			}
 			___cfg80211_scan_done(rdev, false);
 		}
 
-- 
2.34.1

